### ISSUE TITLE
MaxListenersExceededWarning in Vhost.js

### ISSUE THREAD

**Issue by ravihara** [guidesmiths/rascal#164]
Hi, first of all, thanks for this helpful wrapper. I have been using this since almost a year. Of late, I observed the following warning while starting the worker services.

I'm using 13.1.0

```
(node:4824) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 error listeners added to [CallbackModel]. Use emitter.setMaxListeners() to increase limit
    at _addListener (events.js:475:17)
    at CallbackModel.addListener (events.js:497:10)
    at CallbackModel.once (events.js:541:8)
    at createChannel (/home/ubuntu/myproj/node_modules/rascal/lib/amqp/Vhost.js:373:16)
    at createChannelWhenInitialised (/home/ubuntu/myproj/node_modules/rascal/lib/amqp/Vhost.js:346:28)
    at /home/ubuntu/myproj/node_modules/rascal/lib/amqp/Vhost.js:201:11
    at new Promise (<anonymous>)
    at Object.create (/home/ubuntu/myproj/node_modules/rascal/lib/amqp/Vhost.js:199:16)
    at Pool._createResource (/home/ubuntu/myproj/node_modules/generic-pool/lib/Pool.js:319:42)
    at Pool._ensureMinimum (/home/ubuntu/myproj/node_modules/generic-pool/lib/Pool.js:351:12)
    at Pool.start (/home/ubuntu/myproj/node_modules/generic-pool/lib/Pool.js:425:10)
    at new Pool (/home/ubuntu/myproj/node_modules/generic-pool/lib/Pool.js:128:12)
    at Object.createPool (/home/ubuntu/myproj/node_modules/generic-pool/index.js:11:12)
    at createChannelPool (/home/ubuntu/myproj/node_modules/rascal/lib/amqp/Vhost.js:318:24)
    at ensureChannelPools (/home/ubuntu/myproj/node_modules/rascal/lib/amqp/Vhost.js:439:48)
    at /home/ubuntu/myproj/node_modules/rascal/lib/amqp/Vhost.js:63:7
```

Can you please suggest what could be wrong here and any workarounds? Thanks.

**Comment by cressie176** [comment]
Hi @ravihara,

The output comes from Node's event emitter class, which many of the objects in Rascal inherit. It occurs when you register an event listener for the same event more than the maximum number of times (the default is 10). This is a soft limit however, nothing stops working if you continue adding listeners after this.

The reason Node outputs the warning is because it is often caused by a bug in the client code. Why would you want to add 10 error listeners?

In Rascal's case however you may legitimately need to register more than 10 event handlers, especially since connections are shared. Every time you create a channel, Rascal adds [error and close](https://github.com/guidesmiths/rascal/blob/master/lib/amqp/Vhost.js#L372-L373) handlers, so if you have more than 10 channels you could exceed the default. For this reason I [set the max listeners to zero](https://github.com/guidesmiths/rascal/blob/master/lib/amqp/Vhost.js#L395), but this only happens after the channel was successfully created, and if multiple channels are created concurrently maybe it happens to late. 

I'll see if I can reproduce and move the setMaxListeners to earlier in the workflow.

**Comment by ravihara** [comment]
> Hi @ravihara,
> 
> The output comes from Node's event emitter class, which many of the objects in Rascal inherit. It occurs when you register an event listener for the same event more than the maximum number of times (the default is 10). This is a soft limit however, nothing stops working if you continue adding listeners after this.
> 
> The reason Node outputs the warning is because it is often caused by a bug in the client code. Why would you want to add 10 error listeners?
> 
> In Rascal's case however you may legitimately need to register more than 10 event handlers, especially since connections are shared. Every time you create a channel, Rascal adds [error and close](https://github.com/guidesmiths/rascal/blob/master/lib/amqp/Vhost.js#L372-L373) handlers, so if you have more than 10 channels you could exceed the default. For this reason I [set the max listeners to zero](https://github.com/guidesmiths/rascal/blob/master/lib/amqp/Vhost.js#L395), but this only happens after the channel was successfully created, and if multiple channels are created concurrently maybe it happens to late.
> 
> I'll see if I can reproduce and move the setMaxListeners to earlier in the workflow.
Hi @cressie176, thanks for the quick response! Yeah, I was bit reluctant to set MaxListeners to zero, as that is more generally perceived as allowing a probably leaky code. As you mentioned, it is not causing any harm in running our services. The adjustment of moving it early in the workflow might do the trick!! 👍🏽

**Comment by cressie176** [comment]
Fixed in rascal@13.1.1
